core/src/main/java/org/apache/druid/math/expr/Expr.java

```diff
@@ -348,16 +641,30 @@ public void visit(Visitor visitor)
     visitor.visit(this);
   }
 
+  @Override
+  public Expr visit(Shuttle shuttle)
+  {
+    Expr newLeft = left.visit(shuttle);
+    Expr newRight = right.visit(shuttle);
+    if (left != newLeft || right != newRight) {
```

Comment:
this kind of check to avoid copy would be needed if we didn't do defensive copies as suggested before.